libtransmission/peer-mgr.h

```diff
@@ -60,20 +60,28 @@ enum
 class tr_peer_info
 {
 public:
-    tr_peer_info(tr_socket_address socket_address, uint8_t pex_flags, tr_peer_from from)
+    tr_peer_info(
+        tr_socket_address socket_address,
+        uint8_t pex_flags,
```

Comment:
should client_advertised_port be a parameter?  or just knowledge local to the initializer?

Reply:
> or just knowledge local to the initializer?

Could you give a concrete example of what you have in mind (maybe a code snippet)? I'm not sure if I understand you correctly.

> just personal pref that I don't like parameterizing state that so rarely (if ever) changes.

They do change, and as long as we cannot rule out the possibility that it can change, then we need a way to allow the peer info objects to update themselves.

Reply:
> > just personal pref that I don't like parameterizing state that so rarely (if ever) changes.
> 
> They do change, and as long as we cannot rule out the possibility that it can change, then we need a way to allow the peer info objects to update themselves.

When do they change?  I was asserting that the client listen_address and listen_port change very very rarely.  In the majority of cases, they change never.  So why not have this be part of the session state that is reference from within canonical_priority code rather than passing it every time?

as always, apologies if I've misunderstood your code.

Reply:
Many people have dynamic IP addresses instead of static ones. The peer port is also configurable during runtime.

Reply:
Yes, I understand that.  And still, those values change *exceedingly* rarely.  Also, it is always global state, it is never state that is carried per-torrent or per-peer.

Reply:
I'm still not sure how you want the code to be structured, can you maybe show a code snippet?

FYI, I have ruled out saving a `tr_session` ptr/ref in `tr_peer_info`, because this will make it much more complicated than needed to write tests for this class.

Reply:
> I'm still not sure how you want the code to be structured, can you maybe show a code snippet?
> 
> FYI, I have ruled out saving a `tr_session` ptr/ref in `tr_peer_info`, because this will make it much more complicated than needed to write tests for this class.

Never totally understood why tr_peer_info doesn't point back to swarm (which could in term be followed back to torrent and session)? Why is that a bad thing? Is it better as a design pattern to increase the parameterization of methods throughout the codebase rather than carry the same state around within class member pointers?

Also don't understand why this would make tests more difficult, but trust your thinking.

Reply:
The tests will not be harder to write per se, since all the infrastructure for testing with a Tr session is already in place. However, a lot of unrelated stuff would be run as a result of passing a session pointer into `tr_peer_info`, and the test will need to connect to the internet to get an IP address for the session before the canonical priority can be tested, which is not always reliable.